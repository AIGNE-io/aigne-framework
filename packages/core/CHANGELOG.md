## 1.1.0-beta.17 (2025-3-18)

- chore: add support for esm module


## [1.2.0](https://github.com/AIGNE-io/aigne-framework/compare/core-next-v1.1.0...core-next-v1.2.0) (2025-03-18)


### Features

* add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine ([4d2a5a1](https://github.com/AIGNE-io/aigne-framework/commit/4d2a5a1b3366b8f935f50a0937c2da6e49073348))
* add OrchestratorAgent in agent library ([25a5e9e](https://github.com/AIGNE-io/aigne-framework/commit/25a5e9e6c60d747c8bf484ac884b31dc02c14757))
* add sequential and parallel helper function ([a295697](https://github.com/AIGNE-io/aigne-framework/commit/a295697b5694754e02954fc5c7f382a3b219a3ab))
* add support for MCP resources ([1ded2fb](https://github.com/AIGNE-io/aigne-framework/commit/1ded2fbf222fa8984e75df0852ff384524f73b04))
* **prompt-builder:** support chat history in PromptBuilder ([6ca05f2](https://github.com/AIGNE-io/aigne-framework/commit/6ca05f28eddb683a4f1e228865f8bbf8a8e190f1))
* support run puppeteer example chat loop in terminal ([85ce7f8](https://github.com/AIGNE-io/aigne-framework/commit/85ce7f8de8b443c86e50815dd7bcab99f869c4ce))
* use PromptBuilder instead of string instructions ([e4cb2cb](https://github.com/AIGNE-io/aigne-framework/commit/e4cb2cb6baf4f9bcef390567a4a174e9246c29a3))


### Bug Fixes

* **AIAgent:** should pass both arguments (model generated) and input (user input) to tool ([c49d64e](https://github.com/AIGNE-io/aigne-framework/commit/c49d64ee35f7efd83b0f82f43205bb1c40f999e8))
* **MCP:** catch list resource error treat as empty list ([1885fab](https://github.com/AIGNE-io/aigne-framework/commit/1885fab3585e0dd1467b127e5b47cd0b98282bab))
* use text resource from MCP correctly ([8b9eba8](https://github.com/AIGNE-io/aigne-framework/commit/8b9eba83352ec096a2a5d4f410d4c4bde7420bce))

## 1.1.0-beta.16 (2025-3-18)

- chore: add puppeteer in linux need docker_container

## 1.1.0-beta.15 (2025-3-18)

- chore: make coverage report as text to terminal
- chore: update contributing docs

## 1.1.0-beta.14 (2025-3-18)

- chore(example): add code-execution example

## 1.1.0-beta.13 (2025-3-18)

- feat: add OrchestratorAgent in agent library

## 1.1.0-beta.12 (2025-3-14)

- chore(example): add concurrency reflection handoff workflow examples

## 1.1.0-beta.11 (2025-3-14)

- feat(core): add sequential and parallel helper function
- chore(examples): add workflow-sequential example

## 1.1.0-beta.10 (2025-3-13)

- chore: ensure required environment variables have values

## 1.1.0-beta.9 (2025-3-13)

- fix(MCP): catch list resource error treat as empty list

## 1.1.0-beta.8 (2025-3-13)

- fix(AIAgent): should pass both arguments (model generated) and input (user input) to tool
- chore(examples): add workflow-router example
- chore(examples): rename examples puppeteer-mcp-server and sqlite-mcp-server to mcp-server-puppeteer and mcp-server-sqlite

## 1.1.0-beta.7 (2025-3-13)

- chore: rename @aigne/core to @aigne/core-next

## 1.1.0-beta.6 (2025-3-13)

- chore(examples): default enable mcp debug message for examples

## 1.1.0-beta.5 (2025-3-13)

- feat: support chat history in PromptBuilder
- feat: add `prompts` for MCPAgent to consume prompts from MCP server
- chore: add sqlite-mcp-server example
- test: add more unit test cases

## 1.1.0-beta.4 (2025-3-12)

- feat: support run puppeteer example chat loop in terminal

## 1.1.0-beta.3 (2025-3-11)

- chore: set module type for core package

## 1.1.0-beta.2 (2025-3-11)

- feat: use PromptBuilder instead of string instructions
- refactor: use tools instead of skills
- chore(examples): add puppeteer-mcp-server example

## 1.1.0-beta.1 (2025-3-11)

- feat: add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine
