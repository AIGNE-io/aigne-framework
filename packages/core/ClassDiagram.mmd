classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ tools
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    class Model {
        +call(Prompt input) object*
    }

    ChatModel --|> Model: inheritance
    class ChatModel {
    }

    ImageModel --|> Model: inheritance
    class ImageModel {
    }

    Agent --|> EventEmitter: inheritance
    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ inputTopic
        +List~string~ nextTopic
        +List~Agent~ tools
        +Map~string, Agent~ skills

        +call(string input, Context context) object
        +call(object input, Context context) object

        +process(object input, Context context) object*
        -verifyInput() void
        -verifyOutput() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +string instructions
        +string outputKey
        +object toolChoice
        +PromptBuilder promptBuilder
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    class ImageAgent {
        +ImageModel model
        +string instructions
        +PromptBuilder promptBuilder
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function fn
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class History {
        +string id
        +string agentId
        +object input
        +object output
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    Context ..> History: dependency
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ tools

        +getHistories(string agentId, int limit) List~History~
        +addHistory(History history) void
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }

    class ExecutionEngineRunOptions {
        +boolean concurrency
    }

    ExecutionEngine --|> Context: inheritance
    ExecutionEngine --|> EventEmitter: inheritance
    ExecutionEngine ..> UserAgent: dependency
    ExecutionEngine ..> ExecutionEngineRunOptions: dependency
    class ExecutionEngine {
        +run(Agent agent) UserAgent
        +run(string input) object
        +run(object input) object
        +run(object input, Agent ...agents) object
        +run(object input, ExecutionEngineRunOptions options, Agent ...agents) object
    }
