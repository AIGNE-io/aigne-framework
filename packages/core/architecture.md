## Class Definitions

```mermaid
classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ tools
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    class Model {
        +call(Prompt input) object*
    }

    ChatModel --|> Model: inheritance
    class ChatModel {
    }

    ImageModel --|> Model: inheritance
    class ImageModel {
    }

    Agent --|> EventEmitter: inheritance
    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ inputTopic
        +List~string~ nextTopic
        +List~Agent~ tools

        +call(string input, Context context) object
        +call(object input, Context context) object

        +process(object input, Context context) object*
        -verifyInput() void
        -verifyOutput() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +string instructions
        +string outputKey
        +object toolChoice
        +PromptBuilder promptBuilder
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    class ImageAgent {
        +ImageModel model
        +string instructions
        +PromptBuilder promptBuilder
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function fn
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class History {
        +string id
        +string agentId
        +object input
        +object output
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    Context ..> History: dependency
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ tools

        +getHistories(string agentId, int limit) List~History~
        +addHistory(History history) void
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }


    AIGNE --|> Context: inheritance
    AIGNE --|> EventEmitter: inheritance
    AIGNE ..> UserAgent: dependency
    class AIGNE {
        +call(Agent agent) UserAgent
        +call(Agent agent, string input) object
        +call(Agent agent, object input) object
    }
```
