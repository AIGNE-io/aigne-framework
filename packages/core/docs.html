<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport"
    content="width=device-width, initial-scale=1, minimum-scale=1.0, shrink-to-fit=no, viewport-fit=cover">
  <meta name="description" content="">
  <title></title>

  <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/docsify/themes/buble.css" />

  <!-- Custom Styles -->
  <style>
    :root {
      /* --theme-hue: 325; */
    }
  </style>
</head>

<body>
  <div id="app"></div>

  <script>
    // Docsify Configuration
    window.$docsify = {
      name: 'AIGNE Framework',
      loadSidebar: true,
      alias: {
        '/.*/_sidebar.md': '/_sidebar.md'
      },
      relativePath: true
    };
  </script>

  <!-- Required -->
  <script src="https://cdn.jsdelivr.net/npm/docsify@4/lib/docsify.min.js"></script>

  <!-- Recommended -->
  <script src="https://cdn.jsdelivr.net/npm/docsify@4/lib/plugins/search.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/docsify@4/lib/plugins/zoom-image.min.js"></script>

  <script src="//cdn.jsdelivr.net/npm/prismjs@1/components/prism-typescript.min.js"></script>
</body>

</html>
