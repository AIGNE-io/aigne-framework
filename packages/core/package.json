{
  "name": "@aigne/core",
  "version": "1.11.0",
  "description": "AIGNE core library for building AI-powered applications",
  "publishConfig": {
    "access": "public"
  },
  "author": "Arcblock <blocklet@arcblock.io> https://github.com/blocklet",
  "homepage": "https://github.com/AIGNE-io/aigne-framework",
  "license": "Elastic-2.0",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/AIGNE-io/aigne-framework"
  },
  "files": [
    "lib/cjs",
    "lib/dts",
    "lib/esm",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ],
  "type": "module",
  "main": "./lib/cjs/index.js",
  "module": "./lib/esm/index.js",
  "types": "./lib/dts/index.d.ts",
  "exports": {
    ".": {
      "import": "./lib/esm/index.js",
      "require": "./lib/cjs/index.js",
      "types": "./lib/dts/index.d.ts"
    },
    "./*": {
      "import": "./lib/esm/*",
      "require": "./lib/cjs/*",
      "types": "./lib/dts/*"
    }
  },
  "typesVersions": {
    "*": {
      ".": [
        "./lib/dts/index.d.ts"
      ],
      "agents/*": [
        "./lib/dts/agents/*"
      ],
      "execution-engine/*": [
        "./lib/dts/execution-engine/*"
      ],
      "loader/*": [
        "./lib/dts/loader/*"
      ],
      "models/*": [
        "./lib/dts/models/*"
      ],
      "prompt/*": [
        "./lib/dts/prompt/*"
      ],
      "utils/*": [
        "./lib/dts/utils/*"
      ]
    }
  },
  "scripts": {
    "lint": "tsc --noEmit",
    "build": "tsc --build scripts/tsconfig.build.json",
    "clean": "rimraf lib test/coverage",
    "prepublishOnly": "run-s clean build",
    "test": "bun --cwd test test",
    "test:coverage": "bun --cwd test test --coverage --coverage-reporter=lcov --coverage-reporter=text",
    "postbuild": "echo '{\"type\": \"module\"}' > lib/esm/package.json && echo '{\"type\": \"commonjs\"}' > lib/cjs/package.json"
  },
  "dependencies": {
    "@aigne/json-schema-to-zod": "^1.3.3",
    "@modelcontextprotocol/sdk": "^1.9.0",
    "@types/debug": "^4.1.12",
    "camelize-ts": "^3.0.0",
    "debug": "^4.4.0",
    "fast-deep-equal": "^3.1.3",
    "inquirer": "^12.5.2",
    "mustache": "^4.2.0",
    "nanoid": "^5.1.5",
    "p-retry": "^6.2.1",
    "uuid": "^11.1.0",
    "yaml": "^2.7.1",
    "zod": "^3.24.2",
    "zod-to-json-schema": "^3.24.5"
  },
  "peerDependencies": {
    "@anthropic-ai/sdk": "^0.39.0",
    "@google/generative-ai": "^0.24.0",
    "openai": "^4.87.3"
  },
  "devDependencies": {
    "@anthropic-ai/sdk": "^0.39.0",
    "@google/generative-ai": "^0.24.0",
    "@types/bun": "^1.2.9",
    "@types/express": "^5.0.1",
    "@types/mustache": "^4.2.5",
    "@types/node": "^22.14.1",
    "detect-port": "^2.1.0",
    "express": "^5.1.0",
    "npm-run-all": "^4.1.5",
    "openai": "^4.94.0",
    "rimraf": "^6.0.1",
    "typescript": "^5.8.3"
  }
}
