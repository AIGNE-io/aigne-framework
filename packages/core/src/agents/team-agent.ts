import type { Context } from "../aigne/context.js";
import { mergeAgentResponseChunk, readableStreamToAsyncIterator } from "../utils/stream-utils.js";
import { type PromiseOrValue, isEmpty } from "../utils/type-utils.js";
import {
  Agent,
  type AgentOptions,
  type AgentProcessResult,
  type AgentResponseChunk,
  type Message,
} from "./agent.js";

export type TeamAgentProcessMethod = "sequential" | "parallel";

export interface TeamAgentOptions<I extends Message, O extends Message> extends AgentOptions<I, O> {
  /**
   * The method to process the agents in the team.
   * - `sequential`: Process the agents one by one, passing the output of each agent to the next.
   * - `parallel`: Process all agents in parallel, merging the output of all agents.
   * @default "sequential"
   */
  processMethod?: TeamAgentProcessMethod;
}

export class TeamAgent<I extends Message, O extends Message> extends Agent<I, O> {
  static from<I extends Message, O extends Message>(options: TeamAgentOptions<I, O>) {
    return new TeamAgent(options);
  }

  constructor(options: TeamAgentOptions<I, O>) {
    super(options);
    this.processMethod = options.processMethod ?? "sequential";
  }

  processMethod: TeamAgentProcessMethod;

  process(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    switch (this.processMethod) {
      case "sequential":
        return this._processSequential(input, context);
      case "parallel":
        return this._processParallel(input, context);
    }
  }

  async *_processSequential(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    const output: Message = {};

    // Clone the agents to run, so that we can update the agents list during the loop
    const agents = [...this.tools];
    const newAgents: Agent[] = [];

    for (const agent of agents) {
      const [o, transferToAgent] = await context.call(
        agent,
        { ...input, ...output },
        { returnActiveAgent: true, streaming: true },
      );

      for await (const chunk of readableStreamToAsyncIterator(o)) {
        yield chunk;
        mergeAgentResponseChunk(output, chunk);
      }
      newAgents.push(await transferToAgent);
    }

    this.tools.splice(0);
    this.tools.push(...newAgents);
  }

  async *_processParallel(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    const result = await Promise.all(
      this.tools.map((agent) =>
        context.call(agent, input, { returnActiveAgent: true, streaming: true }),
      ),
    );

    const streams = result.map((i) => i[0]);

    type Reader = ReadableStreamDefaultReader<AgentResponseChunk<Message>>;
    type ReaderResult = ReadableStreamReadResult<AgentResponseChunk<Message>>;

    type Task = Promise<{ index: number; reader: Reader } & ReaderResult>;

    const read = async (index: number, reader: Reader): Task => {
      const promise = reader.read();
      return promise.then((result) => ({ ...result, reader, index }));
    };

    const tasks = new Map(streams.map((stream, index) => [index, read(index, stream.getReader())]));

    // NOTE: Flag to check if the output key is used by agent at the index,
    const outputKeyUsed = new Map<string, number>();

    while (tasks.size) {
      const { value, done, reader, index } = await Promise.race(tasks.values());
      tasks.delete(index);
      if (!done) {
        tasks.set(index, read(index, reader));
      }

      if (value) {
        let {
          delta: { text, ...delta },
        } = value;

        if (text) {
          for (const key of Object.keys(text)) {
            // the output key is unused, add to map to lock it
            if (!outputKeyUsed.has(key)) {
              outputKeyUsed.set(key, index);
            }
            // the output key is used by the agent at the index, abandon it
            else if (outputKeyUsed.get(key) !== index) {
              delete text[key];
            }
          }
          if (isEmpty(text)) {
            text = undefined;
          }
        }

        if (!isEmpty(delta.json) || !isEmpty(text)) yield { delta: { ...delta, text } };
      }
    }

    const agents = await Promise.all(result.map((i) => i[1]));
    this.tools.splice(0);
    this.tools.push(...agents);
  }
}
