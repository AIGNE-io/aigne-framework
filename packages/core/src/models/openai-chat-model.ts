import { nanoid } from "nanoid";
import OpenAI from "openai";
import type { ChatCompletionMessageParam, ChatCompletionTool } from "openai/resources";
import { parseJSON } from "../utils/json-schema.js";
import { isNonNullable } from "../utils/type-utils.js";
import {
  ChatModel,
  type ChatModelInput,
  type ChatModelInputMessage,
  type ChatModelInputTool,
  type ChatModelOptions,
  type ChatModelOutput,
  type Role,
} from "./chat-model.js";

const CHAT_MODEL_OPENAI_DEFAULT_MODEL = "gpt-4o-mini";

export interface OpenAIChatModelOptions {
  apiKey?: string;
  baseURL?: string;
  model?: string;
  modelOptions?: ChatModelOptions;
}

export class OpenAIChatModel extends ChatModel {
  constructor(public options?: OpenAIChatModelOptions) {
    super();
  }

  private _client?: OpenAI;

  private get client() {
    if (!this.options?.apiKey) throw new Error("Api Key is required for OpenAIChatModel");

    this._client ??= new OpenAI({
      baseURL: this.options.baseURL,
      apiKey: this.options.apiKey,
    });
    return this._client;
  }

  get modelOptions() {
    return this.options?.modelOptions;
  }

  async process(input: ChatModelInput): Promise<ChatModelOutput> {
    const res = await this.client.chat.completions.create({
      model: this.options?.model || CHAT_MODEL_OPENAI_DEFAULT_MODEL,
      temperature: input.modelOptions?.temperature ?? this.modelOptions?.temperature,
      top_p: input.modelOptions?.topP ?? this.modelOptions?.topP,
      frequency_penalty:
        input.modelOptions?.frequencyPenalty ?? this.modelOptions?.frequencyPenalty,
      presence_penalty: input.modelOptions?.presencePenalty ?? this.modelOptions?.presencePenalty,
      messages: await contentsFromInputMessages(input.messages),
      tools: toolsFromInputTools(input.tools),
      tool_choice: input.toolChoice,
      parallel_tool_calls: !input.tools?.length
        ? undefined
        : (input.modelOptions?.parallelToolCalls ?? this.modelOptions?.parallelToolCalls),
      response_format:
        input.responseFormat?.type === "json_schema"
          ? {
              type: "json_schema",
              json_schema: {
                ...input.responseFormat.jsonSchema,
                schema: jsonSchemaToOpenAIJsonSchema(input.responseFormat.jsonSchema.schema),
              },
            }
          : undefined,
      stream: true,
    });

    let text = "";
    const toolCalls: (NonNullable<ChatModelOutput["toolCalls"]>[number] & {
      args: string;
    })[] = [];

    for await (const chunk of res) {
      const choice = chunk.choices?.[0];

      if (choice?.delta.tool_calls?.length) {
        for (const call of choice.delta.tool_calls) {
          toolCalls[call.index] ??= {
            id: call.id || nanoid(),
            type: "function" as const,
            function: { name: "", arguments: {} },
            args: "",
          };
          const c = toolCalls[call.index];
          if (!c) throw new Error("Tool call not found");

          if (call.type) c.type = call.type;

          c.function.name = c.function.name + (call.function?.name || "");
          c.args = c.args.concat(call.function?.arguments || "");
        }
      }

      if (choice?.delta.content) text += choice.delta.content;
    }

    const result: ChatModelOutput = {};

    if (input.responseFormat?.type === "json_schema" && text) {
      result.json = parseJSON(text);
    } else {
      result.text = text;
    }

    if (toolCalls.length) {
      result.toolCalls = toolCalls.map(({ args, ...c }) => ({
        ...c,
        function: { ...c.function, arguments: parseJSON(args) },
      }));
    }

    return result;
  }
}

const ROLE_MAP: { [key in Role]: ChatCompletionMessageParam["role"] } = {
  system: "system",
  user: "user",
  agent: "assistant",
  tool: "tool",
} as const;

async function contentsFromInputMessages(
  messages: ChatModelInputMessage[],
): Promise<ChatCompletionMessageParam[]> {
  return messages.map(
    (i) =>
      ({
        role: ROLE_MAP[i.role],
        content:
          typeof i.content === "string"
            ? i.content
            : i.content
                ?.map((c) => {
                  if (c.type === "text") {
                    return { type: "text" as const, text: c.text };
                  }
                  if (c.type === "image_url") {
                    return {
                      type: "image_url" as const,
                      image_url: { url: c.url },
                    };
                  }
                })
                .filter(isNonNullable),
        tool_calls: i.toolCalls?.map((i) => ({
          ...i,
          function: {
            ...i.function,
            arguments: JSON.stringify(i.function.arguments),
          },
        })),
        tool_call_id: i.toolCallId,
        name: i.name,
      }) as ChatCompletionMessageParam,
  );
}

function toolsFromInputTools(tools?: ChatModelInputTool[]): ChatCompletionTool[] | undefined {
  return tools?.length
    ? tools.map((i) => ({
        type: "function",
        function: {
          name: i.function.name,
          description: i.function.description,
          parameters: i.function.parameters as Record<string, unknown>,
        },
      }))
    : undefined;
}

function jsonSchemaToOpenAIJsonSchema(schema: Record<string, unknown>): Record<string, unknown> {
  if (schema?.type === "object") {
    const { required, properties } = schema as {
      required?: string[];
      properties: Record<string, unknown>;
    };

    return {
      ...schema,
      properties: Object.fromEntries(
        Object.entries(properties || {}).map(([key, value]) => {
          const valueSchema = jsonSchemaToOpenAIJsonSchema(value as Record<string, unknown>);

          // NOTE: All fields must be required https://platform.openai.com/docs/guides/structured-outputs/all-fields-must-be-required
          return [
            key,
            required?.includes(key) ? valueSchema : { anyOf: [valueSchema, { type: ["null"] }] },
          ];
        }),
      ),
      required: Object.keys(properties || {}),
    };
  }

  if (schema?.type === "array") {
    const { items } = schema as { items: Record<string, unknown> };

    return {
      ...schema,
      items: jsonSchemaToOpenAIJsonSchema(items),
    };
  }

  return schema;
}
