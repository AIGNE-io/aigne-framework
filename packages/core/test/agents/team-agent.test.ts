import { expect, spyOn, test } from "bun:test";
import { AIAgent, AIGNE } from "@aigne/core";
import { ProcessMethod, TeamAgent } from "@aigne/core/agents/team-agent";
import { OpenAIChatModel } from "@aigne/core/models/openai-chat-model.js";
import {
  readableStreamToArray,
  stringToAgentResponseStream,
} from "@aigne/core/utils/stream-utils.js";

const processMethods = Object.values(ProcessMethod);

test.each(processMethods)(
  "TeamAgent should return streaming response with %s process method (multiple agent with different output keys)",
  async (method) => {
    const model = new OpenAIChatModel();

    const aigne = new AIGNE({ model });

    const first = AIAgent.from({
      outputKey: "first",
    });

    const second = AIAgent.from({
      outputKey: "second",
    });

    spyOn(model, "process")
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, ")))
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, world!")));

    const team = TeamAgent.from({
      skills: [first, second],
      processMethod: method,
    });

    const stream = await aigne.invoke(team, "hello", { streaming: true });

    expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
  },
);

test.each(processMethods)(
  "TeamAgent should return streaming response with %s process method (multiple agent with same output key)",
  async (method) => {
    const model = new OpenAIChatModel();

    const aigne = new AIGNE({ model });

    const first = AIAgent.from({
      outputKey: "text",
    });

    const second = AIAgent.from({
      outputKey: "text",
    });

    spyOn(model, "process")
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, ")))
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, world!")));

    const team = TeamAgent.from({
      skills: [first, second],
      processMethod: method,
    });

    const stream = await aigne.invoke(team, "hello", { streaming: true });

    expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
  },
);
