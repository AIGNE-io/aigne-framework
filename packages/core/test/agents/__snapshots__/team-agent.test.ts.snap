// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamAgent should return streaming response with sequential process method (multiple agent with different output keys) 1`] = `
[
  {
    "delta": {
      "text": {
        "first": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "world",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "!",
      },
    },
  },
]
`;

exports[`TeamAgent should return streaming response with parallel process method (multiple agent with different output keys) 1`] = `
[
  {
    "delta": {
      "text": {
        "first": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "world",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "!",
      },
    },
  },
]
`;

exports[`TeamAgent should return streaming response with sequential process method (multiple agent with same output key) 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "world",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "!",
      },
    },
  },
]
`;

exports[`TeamAgent should return streaming response with parallel process method (multiple agent with same output key) 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
]
`;

exports[`TeamAgent with sequential mode should yield output chunks correctly 1`] = `
[
  {
    "delta": {
      "json": {
        "question": "What is AIGNE?",
        "result": [
          {
            "link": "https://example.com/1",
            "title": "First Result",
          },
          {
            "link": "https://example.com/2",
            "title": "Second Result",
          },
        ],
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "First",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "Result",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "Second",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "Result",
      },
    },
  },
]
`;

exports[`TeamAgent with iteratorInputKey should process array input correctly 1`] = `
{
  "sections": [
    {
      "description": "Description for Test title 0",
      "title": "Test title 0",
    },
    {
      "description": "Description for Test title 1",
      "title": "Test title 1",
    },
    {
      "description": "Description for Test title 2",
      "title": "Test title 2",
    },
    {
      "description": "Description for Test title 3",
      "title": "Test title 3",
    },
    {
      "description": "Description for Test title 4",
      "title": "Test title 4",
    },
  ],
}
`;
